crates/vhost-user-backend/src/vring.rs

```diff
@@ -192,6 +196,16 @@ impl<M: GuestAddressSpace> VringState<M> {
         self.queue.set_next_avail(base);
     }
 
+    /// Set the queue next used head index to the value from the used ring.
+    fn fetch_queue_next_used_from_guest(&mut self) -> Result<(), VirtQueError> {
+        let idx = self
+            .queue
+            .used_idx(self.mem.memory().deref(), Ordering::Acquire)?;
```

Comment:
The spec also says: "idx field indicates where the device would put the next descriptor entry in the ring (modulo the queue size). This starts at 0, and increases."

Anyway the Linux driver allocating all memory with __GFP_ZERO resets used and avail to 0 during initialization.
However in case of an error then it resets both avail and used.

So reading it from the guest should be okay.

Reply:
Just checked and `vhost` kernel code does exactly the same, it reads used from the guest memory, but of course in different places:
- vhost-net when VHOST_NET_SET_BACKEND ioctl is received
- vhost-vsock when VHOST_VSOCK_SET_RUNNING ioctl is received

So I think your approach should be fine ;-)